workflow/workflow.go

```diff
@@ -168,8 +167,11 @@ func (e *executableWorkflow) Execute(ctx context.Context, serializedInput any) (
 	if err != nil {
 		return "", nil, fmt.Errorf("bug: cannot obtain input node (%w)", err)
 	}
-	if err := inputNode.Remove(); err != nil {
-		return "", nil, fmt.Errorf("failed to remove input node from DAG (%w)", err)
+	if err := l.dag.PushStartingNodes(); err != nil {
+		return "", nil, fmt.Errorf("failed to setup starting nodes in DAG (%w)", err)
+	}
+	if err := inputNode.ResolveNode(dgraph.Resolved); err != nil {
+		return "", nil, fmt.Errorf("failed to resolve input node in DAG (%w)", err)
 	}
```

Comment:
> I think having just one place for popping ready nodes lowers the chance of missing something accidentally or unnecessarily complicating the code.

I very much agree with that sentiment, which is why it seems weird to be invoking `inputNode.ResolveNode()` here when we have code elsewhere which is prepared to handle this.  (Conversely, it seems weird to have code elsewhere to handle this if it always gets handed here, instead.)

> Here are the nodes returned by the first call to `PopReadyNodes()` in a simple test case

That's _after_ the invocation of `inputNode.ResolveNode()` has already taken place, right?  What would `PopReadyNodes()` return at the point just before invoking `inputNode.ResolveNode()`?  I think it would be just `input`; and, if we called it there, then, at the "normal" call site, `input` wouldn't be in the returned list, and we wouldn't need the special-case code to filter it out...unless there are other cases that that code handles....

Reply:
While I agree that the pop->resolve way of doing things makes sense in most circumstances, I think this is the best place to resolve it. It's allowed to be resolved here, and you can check in the loop to see if the node that you got was resolved or not. Already-resolved nodes have a resolved status, and non-resolved ones are `Waiting`.

Reply:
Other than this super-special case of the primordial input node, do we have any other cases where we resolve nodes without first popping them from the ready list?  If not, then why should we do that here??

> I think this is the best place to resolve it.

That's fine; but, before we resolve it, I think we should remove it from the ready list.  At this point in the code, we know that it is ready and, once we resolve it, it doesn't need to be on the ready list any longer.  And, we _should_ be able to use `PopReadyList()` to do that, because it _should_ be the only node on the list at this point...right?

You said
> I think having just one place for popping ready nodes lowers the chance of missing something accidentally or unnecessarily complicating the code.

I would replace "popping ready nodes" with "resolving nodes" in that statement.

And, I think that you can avoid an unnecessary complication by ensuring that there never are any `Resolved` nodes on the ready list.

Reply:
> The ones that do not follow PopReadyNodes are the input node and the unresolvable nodes

Marking nodes as `Unresolvable` should not be a problem:
- If the target node is `Waiting`:
  - If it is already on the ready list:
    - its inputs have already been satisfied;
    - the current dependency must be an _OR_ dependency; and,
    - the attempt to mark it as `Unresolvable` should have no effect (which is OK).
  - If it is not already on the ready list:
    - if current dependency is an _OR_ dependency, 
      - the attempt to mark it as `Unresolvable` should have no effect (which is OK);
    - if current dependency is an _AND_ dependency,
      - the attempt to mark it as `Unresolvable` should put it on the ready list (which is OK).
- If the target node is `Unresolvable`, marking it is a no-op (which is OK).

In any case, we don't mind (and, in fact, _expect_) that there may be `Unresolvable` nodes on the ready list.

The ones which are marking nodes `Resolved` _should_ be working on nodes that they got via `PopReadyNodes()` (so, the nodes shouldn't be on the ready list after they are marked `Resolved`).

So, really, there is only one case to consider, which is the primordial input node.  If, in that case, we remove it from the ready list before we mark it `Resolved`, then there won't be any `Resolved` nodes on the ready list for subsequent `PopReadyNodes()` calls to return.

Reply:
And I think popping them there could result in accidentally popping something else, if in the future we add another type of node with no dependencies starting node, we could accidentally remove it here.

And it's possible that this includes nodes that have only literals as dependencies. We do not want to pop them here. We cannot selectively pop them.

Reply:
OK, it sounds like we've painted ourselves into a corner.